app/src/main/java/bg/zahov/app/ui/workout/start/StartWorkoutViewModel.kt

```diff
@@ -1,117 +1,229 @@
+ * @property name The name of the workout.
+ * @property date The date when the workout is created.
+ * @property exercises A list of exercise names included in the workout.
+ */
+data class StartWorkout(
+    val id: String,
+    val name: String,
+    val date: String,
+    val exercises: List<String>,
+    val note: String = "",
+    val personalRecords: String = "0"
+)
 
-    private val serviceError by lazy {
-        application.getServiceErrorProvider()
-    }
-    private val _state = MutableLiveData<State>()
-    val state: LiveData<State>
-        get() = _state
+/**
+ * ViewModel responsible for managing the state of the start workout screen.
+ *
+ * @param repo The provider responsible for fetching workout templates.
+ * @param workoutState The state manager responsible for tracking the current workout state.
+ * @param toastManager responsible for showing toasts to the user
+ * @param serviceError The handler for managing service errors.
+ */
+class StartWorkoutViewModel(
+    private val repo: WorkoutProvider = Inject.workoutProvider,
+    private val toastManager: ToastManager = ToastManager,
+    private val workoutState: WorkoutStateManager = Inject.workoutState,
+    private val serviceError: ServiceErrorHandler = Inject.serviceErrorHandler
+) : ViewModel() {
 
-    private val _templates = MutableLiveData<List<Workout>>(listOf())
-    val templates: LiveData<List<Workout>>
-        get() = _templates
+    /**
+     * Holds the current UI state for starting a workout, including the list of available workouts and any user notifications.
+     */
+    private val _uiState = MutableStateFlow<StartWorkoutUiState>(StartWorkoutUiState())
+    val uiState: StateFlow<StartWorkoutUiState> = _uiState
 
+    /**
+     * The current state of the workout [WorkoutState.ACTIVE] or [WorkoutState.INACTIVE] or [WorkoutState.MINIMIZED] .
+     */
+    private var currentWorkoutState: WorkoutState = WorkoutState.ACTIVE
+
+    /**
+     * used for mapping to [Workout] because otherwise we lose information for set types
+     */
+    private var exerciseTemplates: List<Exercise> = listOf()
+
+    /**
+     * collect template workouts, state and exercises which are to be used for mapping
+     */
     init {
-        getWorkouts()
         viewModelScope.launch {
-            workoutState.state.collect {
-                when (it) {
-                    WorkoutState.MINIMIZED -> _state.postValue(State.Active(true))
-                    WorkoutState.INACTIVE -> _state.postValue(State.Active(false))
-                    else -> {}
+            launch {
+                try {
+                    repo.getStartWorkouts().collect { workouts ->
+                        _uiState.update { old ->
+                            old.copy(workouts = workouts)
+                        }
+                    }
+                } catch (e: CriticalDataNullException) {
+                    serviceError.initiateCountdown()
```

Comment:
If we don't show any errors or error screens when an exception is caught here, then these catch blocks can be handled in the View layer instead.

Reply:
👍

Reply:
I think this should be handled in `FirestoreManager` instead of in `WorkoutProviderImpl`. 

Right now a new flow is created when the first flow fails to be obtained.

The idea of flows is to observe a flow of data which pushes results and changes. The error case should can be handled by pushing an error state or just empty data to the same flow.

Reply:
👍

Reply:
Before instead of having an empty constructor for the fields we used to throw whenever the mapper failed to map. Now if that happens we simply provide empty data eliminating the need to throw.